core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1920,7 +1920,8 @@ public void deleteInternal(RpcContext rpcContext, LockedInodePath inodePath,
     }
 
     // Inodes for which deletion will be attempted
-    List<Pair<AlluxioURI, LockedInodePath>> inodesToDelete = new ArrayList<>();
+    List<Pair<AlluxioURI, LockedInodePath>> inodesToDelete =
+        new ArrayList<>((int) inode.asDirectory().getChildCount());
```

Comment:
This is only the direct children count. In order to make a better guess, I will need to heuristically go under this dir and combine children's child counts. Maybe it's worth a try? Just go at least 1 level deeper and combine the counts on children. The cost is most probably smaller than auto-sizing. This optimization is not perfect of course, we can't know without checking.

For this version, at least it helps when this directory is large, reducing memory footprint from `O(2n)` to `O(n)`